pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceTransaction.java

```diff
@@ -0,0 +1,685 @@
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            if (arguments.isEnableTransaction) {
+                printTxnAggregatedThroughput(startTime);
+            } else {
+                printAggregatedThroughput(startTime);
+            }
+            printAggregatedStats();
+        }));
+
+        // start perf test
+        AtomicBoolean executing = new AtomicBoolean(true);
+        new Thread(() -> {
+                 RateLimiter rateLimiter = arguments.openTxnRate > 0
+                    ? RateLimiter.create(arguments.openTxnRate)
+                    : null;
+            for(int i = 0; i < arguments.numTestThreads; i++) {
+                    LongAdder messageSend = new LongAdder();
+                    LongAdder messageReceived = new LongAdder();
+                    executorService.submit(() -> {
+                        //The producer and consumer clients are built in advance, and then this thread is
+                        //responsible for the production and consumption tasks of the transaction through the loop.
+                        //A thread may perform tasks of multiple transactions in a traversing manner.
+                        List<Producer<byte[]>> producers = null;
+                        List<List<Consumer<byte[]>>> consumers = null;
+                        try {
+                            producers = buildProducers(client, arguments);
+                            consumers = buildConsumer(client, arguments);
+                        } catch (Exception e) {
+                            log.error("Failed to build Producer/Consumer with exception : " + e);
+                        }
+                        AtomicReference<Transaction> atomicReference = buildTransaction(client, arguments);
+                        //The while loop has no break, and finally ends the execution through the shutdownNow of
+                        //0the executorService
+                        while (true) {
+                            try {
+                                Transaction transaction = atomicReference.get();
+                                for (List<Consumer<byte[]>> subscriptions : consumers) {
+                                    while(true) {
+                                        if(messageReceived.sum() == arguments.numMessagesReceivedPerTransaction){
+                                            break;
+                                        }
+                                        for (Consumer<byte[]> consumer : subscriptions) {
+                                            if (messageReceived.sum() == arguments.numMessagesReceivedPerTransaction) {
+                                                break;
+                                            }
+                                            Message message = null;
+                                            try {
+                                                message = consumer.receive(2, TimeUnit.SECONDS);
```

Comment:
I think we should not use timeout because it has no effect

Reply:
When tasks were executed completely, `executorService.shutdownNow();` will be executed.  At this time, other threads may be executing, and an interrupt exception may occur